DiamondLightSource/hyperion
# 867 devices in blueapi context

Fixes https://github.com/DiamondLightSource/hyperion/issues/867

This refactors our hyperion plans to no longer refer to things like `i03.backlight()` directly, and instead to pull devices out of a `blueapi` context.

The main approach at present is that each class defines a composite of devices it needs (as a dataclass), in it's setup method.

as an example, something like:

```python
@dataclasses.dataclass
class SomePlanComposite():
    backlight: Backlight
    eiger: Eiger
```

In the plan's `create_devices`, this composite is then instantiated with the appropriate devices from the `blueapi` context. This composite instance is then passed as one of the arguments to the plan, and relevant devices (or the entire composite) passed down into sub-plans as necessary.

I've tried to design this so that it should be relatively easy to move to blueapi's `inject` mechanism when we do https://github.com/DiamondLightSource/hyperion/issues/868 (might need a thin shim), as now all of the dependencies of each plan are passed around as explicit arguments rather than implicit and scattered throughout sub-plans.

I've tried to keep the API type-safe for developers where possible, as before, which is why I chose to use a dataclasses in the above format rather than passing around e.g. a `Dict[str, Device]`. This also lets us check we got a device of the expected type out of the context.

Some API decisions which I'm not necessarily sure about:
- The handling of `wait_for_connection=False` is a bit hacky, it's currently "known about" in two places, one when you setup the context (which honors the wait_for_connection setting) and one when the device composite is actually created (at which point you're starting a plan and therefore always need the devices to be connected).
  * I think this will likely fall out of https://github.com/DiamondLightSource/hyperion/issues/868 once that issue is done, as  `blueapi` will have to be responsible for making sure that any devices it injects are actually connected at the point they're used
  * Unless we want to make the plan itself responsible for this? But that doesn't seem right.
  * Happy to hear any other alternative approaches
- Post-init in `device_instantiation` (for example - loading aperture positions). At the point when we're building the `context` we don't know which plan we'll be running but still need to set up the devices. Currently I've moved loading aperture positions into a `__post_init__` on the dataclasses described above, so that this logic happens when the relevant device is about to be used. 
  * This will likely change again when we do https://github.com/DiamondLightSource/hyperion/issues/868 as `blueapi` will then somehow have to be told about any post-device-creation initialisation that needs to be done.
  * But again, happy to hear alternative approaches on this...

### To test
- Ensure that individual plans do not reference `i03.<anything>` any more.
  * Unit tests might still reference i03, in particular mocks set up using things like `mock_backlight = i03.backlight(fake_with_ophyd_sim=True)`. This seems ok to me as it's just for simulated devices.
- No *new* s03 test failures. Some tests fail on `main` and still fail on this branch (e.g. rotation scans, fgs).
- **Will need testing on beamline(s).** This is a change that touches all plans and isn't possible to test fully from a dev machine as far as I know.
